EA Forum Dark Mode

oetherington opened the pull request:
Fixes #4880 

This PR makes some changes to the way that dark mode works (mainly on the backend) and enables it for non-admins on the EA forum. LessWrong should be uneffected.

The main change is to enable "auto" mode where we choose the theme based off of the user's desktop preferences, and to make this the default for logged-out users and users who haven't made a choice yet. This is stored in the database as `user.theme`, rather than just as a cookie as it currently is on LessWrong.

Theme selection is now also done on the account settings page under "Site Customizations", rather than in the dropdown. I'm not 100% happy with this solution as it seems a bit janky to need to click submit on the form after choosing a new theme rather than it just autosaving, but this seems like a good enough start that we can iterate on (unless anyone has any good ideas now).

The primary mechanism for implementing this on the backend is to split the `ThemeOptions` type into two:
- `ThemeOptions` works as before - it represents a "concrete" theme that corresponds directly to a particular stylesheet (ie; "default" or "dark"), and;
- `AbstractThemeOptions` is a strict superset of `ThemeOptions` allowing "abstract" themes which require some logic to be executed to map to a stylesheet (ie; "auto").

To get this working correctly, we now also include all generated MUI styles in our main `allStyles` stylesheet, instead of embedding them directly in the HTML. For now, this is done by manually tracking which components are used in `server/usedMuiStyles.ts`. There's definitely better solutions which don't require manually updating this, but they also seem to take considerably more effort and, I think, would be best implemented in a separate PR. I don't expect this to cause many issues in the short term, since its pretty rare we actually use a new MUI component that isn't anywhere else in the codebase already.

One other open question is how we should handle preloading stylesheets when in auto mode. As it's currently implemented, we don't load the unused stylesheet until it's actually required, which means the changeover can be quite abrupt when the user's preference changes. The simplest solution to this would be to always just load both sheets, but this will slow down the initial page load as we'd be loading 500kb of unused CSS. Maybe it'd be better to preload the unused sheet after a timeout? I'm interested in others' opinions here.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203225690779584) by [Unito](https://www.unito.io)

jpaddison3 commented:
Working on this is making me so excited

jpaddison3 commented:
> Theme selection is now also done on the account settings page under "Site Customizations", rather than in the dropdown.

Worth noting that LW still has the dropdown menu item. I could give my usual spiel about how the user menu is too crowded, and then @jimrandomh could give his cached response about how its better for users to be able to try out themes so that any day now when he builds a sans-serif theme users can try it, and then when we're done writing our essays @darkruby501 can make his decision about what he wants. @darkruby501 let us know if you want us to regurgitate those arguments, or if you want to keep it in the menu. By default it will stay.

jpaddison3 commented on packages/lesswrong/components/themes/useTheme.tsx:
```diff
@@ -15,29 +21,98 @@ export const useTheme = (): ThemeType => {
   children: React.ReactNode,
 }) => {
-  const [themeOptions,setThemeOptions] = useState(options);
+  const [_cookies, setCookie] = useCookies([THEME_COOKIE_NAME]);
```
I see you setting the cookie, but don't see you reading it anywhere?

jpaddison3 commented:
What do we do with LW users who have cookies set?

Thinking initially about a simple case:

* User has a single laptop. Write migration code that reads the cookie, respects it for that render, writes it to the user object, and then deletes the cookie. (I'm not sure why you're setting cookies still. Maybe you've thought of something I haven't?)

The trickier case is what happens to users who have multiple devices. One situation: 

* A user has a phone where they prefer light mode, but a laptop where they prefer dark mode. They have set their cookies accordingly. Our approach no longer supports this case. We would need to ask @darkruby501 for permission to stop supporting this use-case.

Another case:

* The user has dark mode enabled on both their phone and laptop, but on their laptop, when they read the announcement about the setting, they change to auto. The next day, they visit on their phone. Their phone sends the dark theme cookie, and the server obediently overwrites their setting. We could fix this with a hack like "userHasChangedThemeWithNewThemeOptions" which prevents the second cookie update, but it's annoying, and probably not worth it for like 5 users.

My initial thought it that we should instead store all LW theme preferences in cookies, contingent upon the theme picker staying in the UsersMenu on LW. It solves:
 * Backwards compatibility, for free, including with multiple devices overwriting the user's preference
 * Supporting multiple devices having different preferences, which we ~can't do otherwise

OTOH, if @darkruby501 wants to remove the UsersMenu item, is ok with dropping support for multiple devices having multiple themes, and is ok with the solution to the third bullet point being "those 5 users are confused", then I'd be happy going with a simple migration described in the first bullet point.

I'd leave it to @darkruby501 to make the call.

jpaddison3 commented:
Another LW-affecting comment: I have not heard anything from LW saying that they don't want the theme to respect your system-provided theme preference — only that they thought it was too much work. If it's easy, and I think it probably is, could you add an option to their theme picker for "Auto"?

I will slightly more controversially claim that it should be the default on LW as well as the EA Forum. If the browser is reporting "this users prefers dark themes" I'd say serve them a dark theme by default.

Once again @darkruby501 or @jimrandomh feel free to object.

jpaddison3 commented on packages/lesswrong/components/themes/usePrefersDarkMode.ts:
```diff
@@ -0,0 +1,24 @@
+      removeEventListener: () => {},
+    };
+
+export const usePrefersDarkMode = () => {
+  const [query] = useState(() => buildQuery());
+  const [prefersDarkMode, setPrefersDarkMode] = useState(query.matches);
+
+  useEffect(() => {
+      const handler = ({matches}: MediaQueryListEvent) => setPrefersDarkMode(matches);
+      query.addEventListener("change", handler);
+      return () => query.removeEventListener("change", handler);
+  }, [query]);
+
+  return prefersDarkMode;
+}
```
Fight me: this would be easier to read as a HOC with lifecycle methods

jpaddison3 commented:
I'm happy to admin gate and merge this, but before we deploy to everyone, we need to fix the low contrast between EA blue and the dark background.

jpaddison3 reviewed (commented):
This is all I have for now, I need to go and I have a fair bit of reviewing to do on the way the theme object is passed around. Should be enough for you to chew on.

b0b3rt commented on packages/lesswrong/server/usedMuiStyles.ts:
```diff
@@ -0,0 +1,78 @@
+  };
+
+  // Filter out components that don't have any styles
+  return Object.keys(components).reduce((result, componentName) => {
+    return components[componentName]
+      ? {...result, [componentName]: components[componentName]}
+      : result;
+  }, {});
```
I pretty strongly disprefer reduce when there are nearly-identical constructions, i.e. using `Object.fromEntries(Object.entries(...).filter(...))`, especially since I've got https://github.com/ForumMagnum/ForumMagnum/pull/5840 as a work-in-progress and `reduce` with a starting point of `{}` really doesn't play well with `noImplicitAny: true`.

darkruby501 commented:
@jpaddison3 
My thoughts on the various topics:

- It's a feature I think isn't valuable if it isn't discoverable. For that reason, I like having it in both drop-down and site customization.
- I like the idea of having "auto" as default, seems pretty reasonable, though I think that requires being able to configure it being quite easy to find. Ideally, there's a UI element on every page for it (e.g. the light/dark brightness MUI icon). There's room in the header for it for logged out users, less sure for logged-in users. (The cost of having it come on by default is that it needs to be good, and personally I don't like LW dark mode currently).
- I predict that if we don't support different settings on different devices, then we will get complaints. I'm willing to wait until those complaints come in to do something about it, if you're willing to come back and provide that support when users say they want it.

Also, do I understand correctly now that logged-out users can't save their theme preference if we're not using a cookie?

Historically we've always put things in the user object, but more recently started using cookies to get both (1) multi-device support, (2) provide more convenience to logged-out users. I'd be interested in some generalization code that lets you have user-object variable with cookie-fallback or something

(Another thought that came up in the context of this: would be good to automatically log analytics events on whenever a cookie value is set, since we can't directly read those – if we had that, we'd have more data on current darkmode usage.)

jpaddison3 commented:
@darkruby501 my response is that the default for LW will stay light mode, until you put in that work. We're not building a piece of UI in the header that we're not going to use. Feel free to build a follow up PR.

@oetherington that means I'd like you to make all LW user state continue to be stored in a cookie.

darkruby501 commented:
@jpaddison3 
The UI element isn't a requirement for making it default to auto, just something I think would be good. If you don't want to do it, that's fine. If it's work to make "auto" default for LW, then leaving as is, is fine.

oetherington commented on packages/lesswrong/lib/collections/users/schema.ts:
```diff
@@ -498,11 +498,15 @@ const schema: SchemaType<DbUser> = {
     canRead: ownsOrIsAdmin,
-    hidden: true,
+    hidden: forumTypeSetting.get() !== "EAForum",
+    control: "ThemeSelect",
```
Yeah, that's correct. Originally I added it so that we could preview the theme when the user selects a different option, but I think that's pretty confusing if it doesn't autosave (ie; user selects a different theme, the page updates to show the new theme, user navigates to the home page without submitting the form and the theme reverts to how it was before). I just left the component for now because maybe we want to do something in this space in the future.

WRT the general argument over the UI, I don't really love either of the two options:
 - I 100% agree that the user dropdown menu is too cluttered (plus we've had people saying recently that they think the fact that the username has a dropdown is pretty unintuitive - I can understand Habryka's reasoning here about deterring brand new users from creating new posts, but I think it's perfectly reasonable that brand new users would want to change the theme).
 - Having it in the user settings it problematic because logged-out users can't access it, plus the janky-ness with needing to submit the form/not being able to preview.

Maybe something like this would be better (but prettier, because I just hacked this together in the console...)? It gets it out of the user menu to somewhere a new user might be more likely to look, and it would also be easier to implement autosave and preview.
![Screenshot 2022-10-25 at 13 19 41](https://user-images.githubusercontent.com/5075628/197771398-e1be9354-dcf0-4ad7-b571-333ded336235.png)

jpaddison3 replied:
Let's discuss in 1-1

oetherington commented on packages/lesswrong/components/form-components/ThemeSelect.tsx:
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
```
`error TS2686: 'React' refers to a UMD global, but the current file is a module. Consider adding an import instead`?

jpaddison3 replied:
Er, I meant to refer to the file

oetherington commented on packages/lesswrong/client/themeProvider.tsx:
```diff
@@ -2,11 +2,11 @@ import React from 'react';
 
 
-export function wrapWithMuiTheme (app: React.ReactNode, themeOptions: ThemeOptions) {
+export function wrapWithMuiTheme (app: React.ReactNode, themeOptions: AbstractThemeOptions) {
```
Yep - this is needed. We don't resolve "auto" into a concrete stylesheet until the very last minute.

oetherington commented:
@jpaddison3 I think this is ready for review again now.
 - I've checked it in EAForum mode and LessWrong mode and both seem to be working correctly 
 - We have analytics now
 - Removed the cookie from EAForum
 - Forum gated the colour tweaks
 - Fixed the weird panel background (feel free to overwrite this if you disagree with the solution - I think it's pretty subjective. Replacing `default: shadePalette.grey[20]` with `recentDiscussionThread: shadePalette.grey[0]` would also make a lot of sense)

jpaddison3 commented on packages/lesswrong/components/themes/useTheme.tsx:
```diff
@@ -15,29 +22,112 @@ export const useTheme = (): ThemeType => {
+  document.head.appendChild(styleNode);
+}
+
+const addAutoStylesheet = (id: string, onFinish: OnFinish, siteThemeOverride?: SiteThemeOverride) => {
```
Define auto in a docstring

jpaddison3 commented on packages/lesswrong/components/themes/useTheme.tsx:
```diff
@@ -15,29 +22,112 @@ export const useTheme = (): ThemeType => {
+
   return <ThemeContext.Provider value={themeContext}>
-    <MuiThemeProvider theme={theme as any} sheetsManager={sheetsManager}>
+    <MuiThemeProvider theme={theme} sheetsManager={sheetsManager}>
```
😍

jpaddison3 commented on packages/lesswrong/server/vulcan-lib/apollo-ssr/renderPage.tsx:
```diff
@@ -135,7 +135,7 @@ export const renderWithCache = async (req: Request, res: Response, user: DbUser|
 
-export function getThemeOptionsFromReq(req: Request, user: DbUser|null) {
+export const getThemeOptionsFromReq = (req: Request, user: DbUser|null): AbstractThemeOptions => {
   const themeCookie = getCookieFromReq(req, "theme");
```
JP can't figure something out, that's probably not from this PR:

How does the client know, in javascript, what the theme is? The only time I see us reading from the theme cookie is on the server.

oetherington replied:
We embed the value as `window.themeOptions` during SSR at line 306 of `apolloServer.ts`:
```
const themeOptionsHeader = embedAsGlobalVar("themeOptions", themeOptions);
```

jpaddison3 commented on packages/lesswrong/server/vulcan-lib/apollo-ssr/renderPage.tsx:
```diff
@@ -198,7 +198,14 @@ export const renderRequest = async ({req, user, startTime, res, clientId}: {
+  // aren't entirely obvious to me, this can be fixed by adding <script>0</script> as the
+  // first child of <body> which forces the browser to load the CSS before rendering.
+  // See https://bugzilla.mozilla.org/show_bug.cgi?id=1404468
+  const ssrBody = `<script>0</script><div id="react-app">${htmlContent}</div>`;
```
I don't fully understand the implications of a blocking script before the content, but I'm suss of it. Can you add a condition that only inserts it if the user agent reports firefox?

jpaddison3 commented on packages/lesswrong/server/usedMuiStyles.ts:
```diff
@@ -0,0 +1,74 @@
+  };
+
+  // Filter out components that don't have any styles
+  return Object.fromEntries(Object.entries(components).filter(([_, styles]) => Boolean(styles)));
```
[nit]
 
```suggestion
  return Object.fromEntries(Object.entries(components).filter(([_, styles]) => !!styles));
```

You and Will. I appreciate that there might be arguments for Boolean, but I feel like the codebase style is `!!`.

jpaddison3 commented on packages/lesswrong/server/utils/renderJssSheetImports.ts:
```diff
@@ -0,0 +1,61 @@
+  `<link id="${stylesId}" rel="stylesheet" type="text/css" onerror="window.missingMainStylesheet=true" href="${url}" />`;
+
+const renderImportForColorScheme = (url: string, colorScheme: string): string =>
+  `@import url("${url}") screen and (prefers-color-scheme: ${colorScheme});\n`;
```
I appreciate your newlines

jpaddison3 commented on packages/lesswrong/themes/themeNames.ts:
```diff
@@ -52,25 +71,51 @@ export function isValidSerializedThemeOptions(options: string|object): boolean {
+
+export function getThemeOptions(themeCookie: string | object, user: DbUser|UsersCurrent|null): AbstractThemeOptions {
+  // Check if the user setting is a serialized ThemeOptions object
+  if (user?.theme && isValidSerializedThemeOptions(user.theme)) {
+    return deserializeThemeOptions(user.theme);
+  }
```
I believe there's an issue here where once a LW dev sets their theme while testing the ea forum, they will never be able to change their settings again on localhost.

jpaddison3 replied:
Also, watch out, did LW use to store it in the user object? We're inverting the order of the check.

oetherington replied:
Good point - now we're deleting the token on the EAForum, it's safe to switch the order back for both.

jpaddison3 reviewed (approved):
ASTLAC

Whew. Complicated indeed, but I'm super psyched.

oetherington commented on packages/lesswrong/components/themes/ThemePickerMenu.tsx:
```diff
@@ -29,82 +28,59 @@ const styles = (theme: ThemeType): JssStyles => ({
-          removeStylesheetsMatching(encodeURIComponent(JSON.stringify(oldThemeOptions)));
-        }
+
+  const {mutate: updateUser} = useUpdate({
+    collectionName: "Users",
+    fragmentName: "UsersCurrent",
+  });
```
TIL

oetherington commented:
@jpaddison3 Thanks! Requested changes are done - I'll merge into master on Monday!
